### PR TITLE
Handle visibility archival requests in system workflow

### PR DESCRIPTION
**Pull request by yycptt**
https://github.com/uber/cadence/issues/2564
https://github.com/uber/cadence/issues/2565

### CODE REVIEW

File: service/history/timerQueueProcessorBase.go [hunk 1]
```diff
@@ -420,25 +419,22 @@ func (t *timerQueueProcessorBase) processDeleteHistoryEvent(
...
-			t.metricsClient.IncCounter(metrics.HistoryProcessDeleteHistoryEventScope, metrics.WorkflowCleanupDeleteCount)
-			return t.deleteWorkflow(task, context, msBuilder)
-		}
+	clusterConfiguredForHistoryArchival := t.shard.GetService().GetArchivalMetadata().GetHistoryConfig().ClusterConfiguredForArchival()
+	domainConfiguredForHistoryArchival := domainCacheEntry.GetConfig().HistoryArchivalStatus == workflow.ArchivalStatusEnabled
+	archiveHistory := clusterConfiguredForHistoryArchival && domainConfiguredForHistoryArchival
+
+	clusterConfiguredForVisibilityArchival := t.shard.GetService().GetArchivalMetadata().GetVisibilityConfig().ClusterConfiguredForArchival()
+	domainConfiguredForVisibilityArchival := domainCacheEntry.GetConfig().VisibilityArchivalStatus == workflow.ArchivalStatusEnabled
+	archiveVisibility := clusterConfiguredForVisibilityArchival && domainConfiguredForVisibilityArchival
```
Comment (andrewjdawson2016) [inline service/history/timerQueueProcessorBase.go]: Nice :)

File: service/history/timerQueueProcessorBase.go [hunk 2]
```diff
@@ -508,8 +532,10 @@ func (t *timerQueueProcessorBase) archiveWorkflow(
 	if err := t.deleteWorkflowExecution(task); err != nil {
 		return err
 	}
-	if resp.ArchivedInline {
-		t.metricsClient.IncCounter(metrics.HistoryProcessDeleteHistoryEventScope, metrics.WorkflowCleanupDeleteHistoryInlineCount)
+	if !archiveHistory || resp.HistoryArchivedInline {
```
Comment (andrewjdawson2016) [inline service/history/timerQueueProcessorBase.go]: Can you include a comment explaining that visibility delete happens differently from history and thats why we do not see this same bit of code for visibility?

File: service/worker/archiver/activities.go [hunk 3]
```diff
@@ -136,3 +139,52 @@ func deleteHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
 	}
 	return err
 }
+
+func archiveVisibilityActivity(ctx context.Context, request ArchiveRequest) (err error) {
```
Comment (andrewjdawson2016) [inline service/worker/archiver/activities.go]: This is great how simple this approach is :)

File: service/worker/archiver/client.go [hunk 4]
```diff
@@ -49,21 +50,37 @@ type (
...
+		ExecutionTimestamp int64
+		CloseTimestamp     int64
+		CloseStatus        shared.WorkflowExecutionCloseStatus
+		HistoryLength      int64
+		Memo               *shared.Memo
+		SearchAttributes   map[string][]byte
+		VisibilityURI      string
+
+		// archival targets: history and/or visibility
+		Targets []string
```
Comment (andrewjdawson2016) [inline service/worker/archiver/client.go]: nit: lets make this a list of enums

File: service/worker/archiver/workflow_test.go [hunk 5]
```diff
@@ -132,6 +132,19 @@ func (s *workflowSuite) TestArchivalWorkflow_Success() {
 	env.AssertExpectations(s.T())
 }
 
+// func (s *workflowSuite) TestReplayArchiveHistoryWorkflow() {
```
Comment (andrewjdawson2016) [inline service/worker/archiver/workflow_test.go]: Why is this commented out?

File: service/history/timerQueueProcessorBase.go (hunk 2 above)
Comment (yycptt) [reply service/history/timerQueueProcessorBase.go]: Done.

File: service/worker/archiver/client.go (hunk 4 above)
Comment (yycptt) [reply service/worker/archiver/client.go]: Done.

File: service/worker/archiver/workflow_test.go (hunk 5 above)
Comment (yycptt) [reply service/worker/archiver/workflow_test.go]: Discussed offline.

**Pull request merged**